sdk/webpubsub/azure-webpubsub-client/azure/webpubsub/client/_client.py

```diff
@@ -0,0 +1,776 @@
+                if message.sequence_id:
+                    if not self._sequence_id.try_update(message.sequence_id):
+                        # // drop duplicated message
+                        return
+
+                self._call_back(
+                    CallBackType.SERVER_MESSAGE,
+                    OnServerDataMessageArgs(
+                        data_type=message.data_type, data=message.data, sequence_id=message.sequence_id
+                    ),
+                )
+
+            parsed_message = self._protocol.parse_messages(data)
+            if parsed_message is None:
+                #  None means the message is not recognized.
+                return
+            if parsed_message.kind == "connected":
+                handle_connected_message(parsed_message)
+            elif parsed_message.kind == "disconnected":
+                handle_disconnected_message(parsed_message)
+            elif parsed_message.kind == "ack":
+                handle_ack_message(parsed_message)
+            elif parsed_message.kind == "groupData":
+                handle_group_data_message(parsed_message)
+            elif parsed_message.kind == "serverData":
+                handle_server_data_message(parsed_message)
+            else:
+                _LOGGER.warning("unknown message type: {}", parsed_message.kind)
+
+        def on_close(_: Any, close_status_code: int, close_msg: str):
+            if self._state == WebPubSubClientState.CONNECTED:
+                _LOGGER.info("WebSocket connection closed. Code: %s, Reason: %s", close_status_code, close_msg)
+
+                self._last_close_event = CloseEvent(close_status_code=close_status_code, close_reason=close_msg)
+                # clean ack cache
+                self._ack_map.clear()
+
+                if self._is_stopping:
+                    _LOGGER.warning("The client is stopping state. Stop recovery.")
+                    self._handle_connection_close_and_no_recovery()
+                    return
+
+                if self._last_close_event and self._last_close_event.close_status_code == 1008:
+                    _LOGGER.warning("The websocket close with status code 1008. Stop recovery.")
+                    self._handle_connection_close_and_no_recovery()
+                    return
+
+                if not self._protocol.is_reliable_sub_protocol:
+                    _LOGGER.warning("The protocol is not reliable, recovery is not applicable")
+                    self._handle_connection_close_and_no_recovery()
+                    return
+
+                recovery_url = self._build_recovery_url()
+                if not recovery_url:
+                    _LOGGER.warning("Connection id or reconnection token is not available")
+                    self._handle_connection_close_and_no_recovery()
+                    return
+
+                self._state = WebPubSubClientState.RECOVERING
+                i = 0
+                while i < 30 or self._is_stopping:
```

Comment:
is the value for retries to connect something that can be brought in from a retry policy ?

Reply:
It is related with [connection-recovery](https://github.com/Azure/azure-webpubsub/blob/main/protocols/client/client-spec.md#14-connection-recovery) which is different with message_retry_policy and reconnect_retry_policy. In one word, service will keep data for 30 seconds if the client reconnects after drop accidently. To make it easier to maintain, I update with internal global variables instead of number.